core/executor/src/system_contract/metadata/store.rs

```diff
@@ -155,4 +157,63 @@ impl MetadataStore {
     fn get_epoch_by_block_number(&self, block_number: u64) -> ProtocolResult<u64> {
         self.get_epoch_segment()?.get_epoch_number(block_number)
     }
+
+    pub fn set_hardfork_info(&mut self, block_number: u64, info: U256) -> ProtocolResult<()> {
+        let current_info = {
+            match self.trie.get(HARDFORK_KEY.as_bytes())? {
+                Some(data) => {
+                    let mut hardfork_info = HardforkInfo::decode(data)?;
+                    if let Some(i) = hardfork_info.inner.last_mut() {
+                        let info = i.flags & info;
+                        if i.block_number == block_number {
+                            i.flags = info;
+                        } else {
+                            hardfork_info.inner.push(HardforkInfoInner {
+                                block_number,
+                                flags: info,
+                            });
+                        }
```

Comment:
Abstract the code as a `fn add(&self, other: HardforkInfoInner)` method of `HardforkInfo`.